File: ureka-framework-main/ureka_framework/logic/pipeline_flow/flow_issue_u_token.py
# Deployment Environment
from ureka_framework.environment import Environment

# Data Model (RAM)
from ureka_framework.model.shared_data import SharedData
import ureka_framework.model.data_model.this_device as this_device

# Data Model (Message)
import ureka_framework.model.message_model.message as message
import ureka_framework.model.message_model.u_ticket as u_ticket
from ureka_framework.model.message_model.u_ticket import UTicket
import ureka_framework.model.message_model.r_ticket as r_ticket
from ureka_framework.model.message_model.r_ticket import RTicket

# Resource (Logger)
from ureka_framework.resource.logger.simple_logger import simple_log

# Measure Helper
from ureka_framework.logic.stage_worker.measure_helper import MeasureHelper

# Stage Worker
from ureka_framework.logic.stage_worker.received_msg_storer import ReceivedMsgStorer
from ureka_framework.logic.stage_worker.msg_verifier import MsgVerifier
from ureka_framework.logic.stage_worker.executor import Executor
from ureka_framework.logic.stage_worker.msg_generator import MsgGenerator
from ureka_framework.logic.stage_worker.generated_msg_storer import GeneratedMsgStorer
from ureka_framework.logic.stage_worker.msg_sender import MsgSender

# Pipeline Flow
from ureka_framework.logic.pipeline_flow.flow_apply_u_ticket import FlowApplyUTicket


class FlowIssueUToken:
    def __init__(
        self,
        share_data: SharedData,
        measure_helper: MeasureHelper,
        received_msg_storer: ReceivedMsgStorer,
        msg_verifier: MsgVerifier,
        executor: Executor,
        msg_generator: MsgGenerator,
        generated_msg_storer: GeneratedMsgStorer,
        msg_sender: MsgSender,
        flow_apply_u_ticket: FlowApplyUTicket,
    ) -> None:
        self.shared_data = share_data
        self.measure_helper = measure_helper
        self.received_msg_storer = received_msg_storer
        self.msg_verifier = msg_verifier
        self.executor = executor
        self.msg_generator = msg_generator
        self.generated_msg_storer = generated_msg_storer
        self.msg_sender = msg_sender
        self.flow_apply_u_ticket = flow_apply_u_ticket

    ######################################################
    # [PIPELINE FLOW]
    #
    # APY (PS):
    #       holder_send_cmd()   -> _device_recv_cmd()
    #       _holder_recv_data() <- _device_send_data()
    #                           ...
    #                       ..repeated..
    #                           ...
    #       holder_send_cmd(access_end) -> _device_recv_cmd(access_end)
    #       _holder_recv_r_ticket() <- _device_send_r_ticket()
    ######################################################
    def holder_send_cmd(
        self, device_id: str, cmd: str, access_end: bool = False
    ) -> None:
        ######################################################
        # Start Process Measurement
        ######################################################
        self.measure_helper.measure_process_perf_start()

        try:
            # [STAGE: (VL)]
            if device_id in self.shared_data.device_table:
                # [STAGE: (E)]
                self.executor._execute_ps(executing_case="send-utoken", plaintext=cmd)

                if access_end == False:
                    # [STAGE: (C)]
                    self.executor._change_state(this_device.STATE_AGENT_WAIT_FOR_DATA)
                    # [STAGE: (G)]
                    u_ticket_type = f"{u_ticket.TYPE_CMD_UTOKEN}"
                else:
                    # [STAGE: (C)]
                    self.executor._change_state(this_device.STATE_AGENT_WAIT_FOR_RT)
                    # [STAGE: (G)]
                    u_ticket_type = f"{u_ticket.TYPE_ACCESS_END_UTOKEN}"

                # [STAGE: (G)]
                generated_request: dict = {
                    "device_id": f"{self.shared_data.current_session.current_device_id}",
                    "u_ticket_type": f"{u_ticket_type}",
                    "associated_plaintext_cmd": f"{self.shared_data.current_session.associated_plaintext_cmd}",
                    "ciphertext_cmd": f"{self.shared_data.current_session.ciphertext_cmd}",
                    "gcm_authentication_tag_cmd": f"{self.shared_data.current_session.gcm_authentication_tag_cmd}",
                    "iv_data": f"{self.shared_data.current_session.iv_data}",
                }
                generated_u_ticket_json: str = (
                    self.msg_generator._generate_xxx_u_ticket(generated_request)
                )
                # simple_log("debug", f"Generated UToken: {generated_u_ticket_json}")

                ######################################################
                # End Process Measurement
                ######################################################
                if access_end == False:
                    self.measure_helper.measure_recv_cli_perf_time("holder_send_cmd")
                else:
                    self.measure_helper.measure_recv_cli_perf_time("holder_send_access_end_cmd")

                ########################################################################
                # Start Comm Measurement
                ########################################################################
                if self.shared_data.this_device.device_name != "iot_device":
                    self.measure_helper.measure_comm_perf_start()

                # [STAGE: (S)]
                self.msg_sender._send_xxx_message(
                    message.MESSAGE_VERIFY_AND_EXECUTE,
                    u_ticket.MESSAGE_TYPE,
                    generated_u_ticket_json,
                )

        except KeyError:  # pragma: no cover -> FAILURE: (VL)
            error = "FAILURE: (VL)"
            self.shared_data.result_message = f"{error}"
            raise RuntimeError(f"{error}")

        except RuntimeError:  # pragma: no cover -> Weird TK-Request (ValidationError)
            failure_msg = f"FAILURE: (VTKREQ)"
            simple_log("error", failure_msg)

        except:  # pragma: no cover -> Shouldn't Reach Here
            raise RuntimeError(f"Shouldn't Reach Here")

    def _device_recv_cmd(self, received_u_token: UTicket) -> None:
        try:
            # [STAGE: (R)(VR)]

            # [STAGE: (VUT)]
            self.msg_verifier.verify_u_ticket_can_execute(received_u_token)

            # [STAGE: (VTK)(VTS)]
            # [STAGE: (E)]
            self.executor._execute_xxx_u_ticket(received_u_token)
            self.shared_data.result_message = f"-> SUCCESS: VERIFY_UT_CAN_EXECUTE"

            if received_u_token.u_ticket_type == u_ticket.TYPE_CMD_UTOKEN:
                # [STAGE: (C)]
                self.executor._change_state(this_device.STATE_DEVICE_WAIT_FOR_CMD)
            elif received_u_token.u_ticket_type == u_ticket.TYPE_ACCESS_END_UTOKEN:
                # [STAGE: (C)]
                self.executor._change_state(this_device.STATE_DEVICE_WAIT_FOR_UT)
            else:  # pragma: no cover -> Shouldn't Reach Here
                raise RuntimeError(f"Shouldn't Reach Here")

        except RuntimeError as error:
            self.shared_data.result_message = f"{error}"

            # [STAGE: (C)]
            self.executor._change_state(this_device.STATE_DEVICE_WAIT_FOR_CMD)

        except:  # pragma: no cover -> Shouldn't Reach Here
            raise RuntimeError(f"Shouldn't Reach Here")

        finally:
            # PS
            if received_u_token.u_ticket_type == u_ticket.TYPE_CMD_UTOKEN:
                # [STAGE: (G)(S)]
                self._device_send_data(self.shared_data.result_message)
            elif received_u_token.u_ticket_type == u_ticket.TYPE_ACCESS_END_UTOKEN:
                # [STAGE: (G)(S)]
                self.flow_apply_u_ticket._device_send_r_ticket(
                    received_u_token.u_ticket_type,
                    received_u_token.u_ticket_id,
                    self.shared_data.result_message,
                )
            else:  # pragma: no cover -> Shouldn't Reach Here
                raise RuntimeError(f"Shouldn't Reach Here")

    def _device_send_data(self, result_message: str) -> None:
        try:
            # [STAGE: (G)]
            if "SUCCESS" in result_message:
                generated_request: dict = {
                    "r_ticket_type": f"{r_ticket.TYPE_DATA_RTOKEN}",
                    "device_id": f"{self.shared_data.this_device.device_pub_key_str}",
                    "result": f"{result_message}",
                    "audit_start": f"{self.shared_data.current_session.current_u_ticket_id}",
                    "associated_plaintext_data": f"{self.shared_data.current_session.associated_plaintext_data}",
                    "ciphertext_data": f"{self.shared_data.current_session.ciphertext_data}",
                    "gcm_authentication_tag_data": f"{self.shared_data.current_session.gcm_authentication_tag_data}",
                    "iv_cmd": f"{self.shared_data.current_session.iv_cmd}",
                }
            else:
                generated_request: dict = {
                    "r_ticket_type": f"{r_ticket.TYPE_DATA_RTOKEN}",
                    "device_id": f"{self.shared_data.this_device.device_pub_key_str}",
                    "result": f"{result_message}",
                }
            generated_r_ticket_json: str = self.msg_generator._generate_xxx_r_ticket(
                generated_request
            )
            # simple_log("debug", f"Generated RToken: {generated_r_ticket_json}")

            ######################################################
            # End Process Measurement
            ######################################################
            self.measure_helper.measure_recv_msg_perf_time(
                "_device_recv_cmd_and_send_data"
            )

            ########################################################################
            # Start Comm Measurement
            ########################################################################
            if self.shared_data.this_device.device_name != "iot_device":
                self.measure_helper.measure_comm_perf_start()

            # [STAGE: (S)]
            self.msg_sender._send_xxx_message(
                message.MESSAGE_VERIFY_AND_EXECUTE,
                r_ticket.MESSAGE_TYPE,
                generated_r_ticket_json,
            )

        except:  # pragma: no cover -> Shouldn't Reach Here
            raise RuntimeError(f"Shouldn't Reach Here")

        # Manually Finish Simulated Comm
        simple_log(
            "debug",
            f"+ {self.shared_data.this_device.device_name} manually finish PS~~ (device)",
        )
        if Environment.COMMUNICATION_CHANNEL == "SIMULATED":
            self.msg_sender.complete_simulated_comm()

    def _holder_recv_data(self, received_r_token: RTicket) -> None:
        try:
            # [STAGE: (R)(VR)]

            # Query Corresponding UTicket
            # [STAGE: (VL)(L)]
            device_id = received_r_token.device_id
            stored_u_ticket_json = self.shared_data.device_table[
                device_id
            ].device_u_ticket_for_owner
            simple_log("debug", f"Corresponding UTicket: {stored_u_ticket_json}")
            # [STAGE: (VR)]
            stored_u_ticket = self.msg_verifier._classify_u_ticket_is_defined_type(
                stored_u_ticket_json
            )

            # [STAGE: (VRT)]
            self.msg_verifier.verify_u_ticket_has_executed_through_r_ticket(
                r_ticket_in=received_r_token,
                audit_start_ticket=stored_u_ticket,
                audit_end_ticket=None,
            )

            # [STAGE: (VTK)]
            # [STAGE: (E)]
            self.executor._execute_xxx_r_ticket(
                r_ticket_in=received_r_token, comm_end="holder-or-device"
            )
            self.shared_data.result_message = f"-> SUCCESS: VERIFY_UT_HAS_EXECUTED"

            # [STAGE: (C)]
            self.executor._change_state(this_device.STATE_DEVICE_WAIT_FOR_CMD)

            simple_log("debug", f"result_message = {self.shared_data.result_message}")

        except KeyError:  # pragma: no cover -> FAILURE: (VL)
            error = "FAILURE: (VL)"
            self.shared_data.result_message = f"{error}"
            raise RuntimeError(f"{error}")

        except RuntimeError as error:  # FAILURE: (VR)(VRT)(VTK)
            self.shared_data.result_message = f"{error}"

        except:  # pragma: no cover -> Shouldn't Reach Here
            raise RuntimeError(f"Shouldn't Reach Here")

        finally:
            ######################################################
            # End Process Measurement
            ######################################################
            self.measure_helper.measure_recv_msg_perf_time("_holder_recv_data")

            ########################################################################
            # Start Comm Measurement
            ########################################################################
            if self.shared_data.this_device.device_name != "iot_device":
                self.measure_helper.measure_comm_perf_start()

        # Manually Finish Simulated/Bluetooth Comm
        simple_log(
            "debug",
            f"+ {self.shared_data.this_device.device_name} manually finish PS~~ (holder)",
        )
        if Environment.COMMUNICATION_CHANNEL == "SIMULATED":
            self.msg_sender.complete_simulated_comm()
        elif Environment.COMMUNICATION_CHANNEL == "BLUETOOTH":
            self.msg_sender.complete_bluetooth_comm()

File: UrekaApp/app/src/main/java/ureka/framework/logic/pipeline_flow/FlowIssueUToken.java
package ureka.framework.logic.pipeline_flow;

import java.util.HashMap;
import java.util.Map;
import java.util.NoSuchElementException;

import ureka.framework.Environment;
import ureka.framework.logic.stage_worker.Executor;
import ureka.framework.logic.stage_worker.GeneratedMsgStorer;
import ureka.framework.logic.stage_worker.MeasureHelper;
import ureka.framework.logic.stage_worker.MsgGenerator;
import ureka.framework.logic.stage_worker.MsgSender;
import ureka.framework.logic.stage_worker.MsgVerifier;
import ureka.framework.logic.stage_worker.ReceivedMsgStorer;
import ureka.framework.model.SharedData;
import ureka.framework.model.data_model.ThisDevice;
import ureka.framework.model.message_model.Message;
import ureka.framework.model.message_model.RTicket;
import ureka.framework.model.message_model.UTicket;
import ureka.framework.resource.logger.SimpleLogger;

public class FlowIssueUToken {
    private SharedData sharedData;
    private MeasureHelper measureHelper;
    private ReceivedMsgStorer receivedMsgStorer;
    private MsgVerifier msgVerifier;
    private Executor executor;
    private MsgGenerator msgGenerator;
    private GeneratedMsgStorer generatedMsgStorer;
    private MsgSender msgSender;
    private FlowApplyUTicket flowApplyUTicket;

    public FlowIssueUToken(
            SharedData shareData,
            MeasureHelper measureHelper,
            ReceivedMsgStorer receivedMsgStorer,
            MsgVerifier msgVerifier,
            Executor executor,
            MsgGenerator msgGenerator,
            GeneratedMsgStorer generatedMsgStorer,
            MsgSender msgSender,
            FlowApplyUTicket flowApplyUTicket
    ) {
        this.sharedData = shareData;
        this.measureHelper = measureHelper;
        this.receivedMsgStorer = receivedMsgStorer;
        this.msgVerifier = msgVerifier;
        this.executor = executor;
        this.msgGenerator = msgGenerator;
        this.generatedMsgStorer = generatedMsgStorer;
        this.msgSender = msgSender;
        this.flowApplyUTicket = flowApplyUTicket;
    }

    public void setSharedData(SharedData sharedData) {
        this.sharedData = sharedData;
    }

    public SharedData getSharedData() {
        return sharedData;
    }

    public void setMeasureHelper(MeasureHelper measureHelper) {
        this.measureHelper = measureHelper;
    }

    public MeasureHelper getMeasureHelper() {
        return measureHelper;
    }

    public void setReceivedMsgStorer(ReceivedMsgStorer receivedMsgStorer) {
        this.receivedMsgStorer = receivedMsgStorer;
    }

    public ReceivedMsgStorer getReceivedMsgStorer() {
        return receivedMsgStorer;
    }

    public void setMsgVerifier(MsgVerifier msgVerifier) {
        this.msgVerifier = msgVerifier;
    }

    public MsgVerifier getMsgVerifier() {
        return msgVerifier;
    }

    public void setExecutor(Executor executor) {
        this.executor = executor;
    }

    public Executor getExecutor() {
        return executor;
    }

    public void setMsgGenerator(MsgGenerator msgGenerator) {
        this.msgGenerator = msgGenerator;
    }

    public MsgGenerator getMsgGenerator() {
        return msgGenerator;
    }

    public void setGeneratedMsgStorer(GeneratedMsgStorer generatedMsgStorer) {
        this.generatedMsgStorer = generatedMsgStorer;
    }

    public GeneratedMsgStorer getGeneratedMsgStorer() {
        return generatedMsgStorer;
    }

    public void setMsgSender(MsgSender msgSender) {
        this.msgSender = msgSender;
    }

    public MsgSender getMsgSender() {
        return msgSender;
    }

    public void setFlowApplyUTicket(FlowApplyUTicket flowApplyUTicket) {
        this.flowApplyUTicket = flowApplyUTicket;
    }

    public FlowApplyUTicket getFlowApplyUTicket() {
        return flowApplyUTicket;
    }

    /* [PIPELINE FLOW]

     APY (PS):
           holderSendCmd()   -> _deviceRecvCmd()
           _holderRecvData() <- _deviceSendData()
                               ...
                           ..repeated..
                               ...
           holderSendCmd(ACCESS_END) -> _deviceRecvCmd(ACCESS_END)
           _holderRecvRTicket() <- _deviceSendRTicket()
     */
    public void holderSendCmd(String device_id, String cmd, boolean accessEnd) {
        // Start Process Measurement
        this.measureHelper.measureProcessPerfStart();

        try {
            // Stage: (VL)
            if (this.sharedData.getDeviceTable().containsKey(device_id)) {
                // Stage: (E)
                this.executor.executePs("sendUToken", null, cmd, null);

                String u_ticket_type;
                if (!accessEnd) {
                    // Stage: (C)
                    this.executor.changeState(ThisDevice.STATE_AGENT_WAIT_FOR_DATA);
                    // Stage: (G)
                    u_ticket_type = UTicket.TYPE_CMD_UTOKEN;
                } else {
                    // Stage: (C)
                    this.executor.changeState(ThisDevice.STATE_AGENT_WAIT_FOR_RT);
                    // Stage: (G)
                    u_ticket_type = UTicket.TYPE_ACCESS_END_UTOKEN;
                }

                // Stage: (G)
                Map<String, String> generatedRequest = new HashMap<>();
                generatedRequest.put("device_id", this.sharedData.getCurrentSession().getCurrentDeviceId());
                generatedRequest.put("u_ticket_type", u_ticket_type);
                generatedRequest.put("associated_plaintext_cmd", this.sharedData.getCurrentSession().getAssociatedPlaintextCmd());
                generatedRequest.put("ciphertext_cmd", this.sharedData.getCurrentSession().getCiphertextCmd());
                generatedRequest.put("gcm_authentication_tag_cmd", this.sharedData.getCurrentSession().getGcmAuthenticationTagCmd());
                generatedRequest.put("iv_data", this.sharedData.getCurrentSession().getIvData());

                String generatedUTicketJson = this.msgGenerator.generateXxxUTicket(generatedRequest);

                // End Process Measurement
                if (!accessEnd) {
                    this.measureHelper.measureRecvCliPerfTime("holderSendCmd");
                } else {
                    this.measureHelper.measureRecvCliPerfTime("holderSendAccessEndCmd");
                }

                // Start Comm Measurement
                if (!this.sharedData.getThisDevice().getDeviceName().equals("iotDevice")) {
                    this.measureHelper.measureCommPerfStart();
                }

                // Stage: (S)
                this.msgSender.sendXxxMessage(Message.MESSAGE_VERIFY_AND_EXECUTE, UTicket.MESSAGE_TYPE, generatedUTicketJson);

            } else {
                throw new NoSuchElementException("Device ID not found in device table: " + device_id);
            }

        } catch (NoSuchElementException e) {
            String error = "FAILURE: (VL)";
            this.sharedData.setResultMessage(error);
            throw new RuntimeException(error, e);
        } catch (RuntimeException e) {
            String failureMsg = "FAILURE: (VTKREQ)";
            SimpleLogger.simpleLog("error", failureMsg);
            this.sharedData.setResultMessage(failureMsg);
        } catch (Exception e) {
            throw new RuntimeException("Shouldn't Reach Here");
        }
    }

    public void _deviceRecvCmd(UTicket receivedUToken) {
        try {
            // [STAGE: (R)(VR)]

            // [STAGE: (VUT)]
            this.msgVerifier.verifyUTicketCanExecute(receivedUToken);

            // [STAGE: (VTK)(VTS)]
            // [STAGE: (E)]
            this.executor.executeXxxUTicket(receivedUToken);
            this.sharedData.setResultMessage("-> SUCCESS: VERIFY_UT_CAN_EXECUTE");

            if(receivedUToken.getUTicketType().equals(UTicket.TYPE_CMD_UTOKEN)) {
                // [STAGE: (C)]
                this.executor.changeState(ThisDevice.STATE_DEVICE_WAIT_FOR_CMD);
            } else if (receivedUToken.getUTicketType().equals(UTicket.TYPE_ACCESS_END_UTOKEN)) {
                // [STAGE: (C)]
                this.executor.changeState(ThisDevice.STATE_DEVICE_WAIT_FOR_UT);
            } else { // pragma: no cover -> Shouldn't Reach Here
                throw new RuntimeException("Shouldn't Reach Here");
            }
        } catch (RuntimeException error) {
            this.sharedData.setResultMessage(error.getMessage());
            // [STAGE: (C)]
            this.executor.changeState(ThisDevice.STATE_DEVICE_WAIT_FOR_CMD);
        } catch (Exception e) { // pragma: no cover -> Shouldn't Reach Here
            throw new RuntimeException("Shouldn't Reach Here");
        } finally {
            // PS
            if (receivedUToken.getUTicketType().equals(UTicket.TYPE_CMD_UTOKEN)) {
                // [STAGE: (G)(S)]
                this._deviceSendData(this.sharedData.getResultMessage());
            } else if (receivedUToken.getUTicketType().equals(UTicket.TYPE_ACCESS_END_UTOKEN)) {
                // [STAGE: (G)(S)]
                this.flowApplyUTicket._deviceSendRTicket(
                        receivedUToken.getUTicketType(),
                        receivedUToken.getUTicketId(),
                        this.sharedData.getResultMessage()
                );
            } else {
                throw new RuntimeException("Shouldn't Reach Here");
            }
        }
    }

    private void _deviceSendData(String resultMessage) {
        try {
            Map<String, String> generatedRequest = new HashMap<>();
            // // [STAGE: (G)]
            if (resultMessage.contains("SUCCESS")) {
                generatedRequest.put("r_ticket_type", RTicket.TYPE_DATA_RTOKEN);
                generatedRequest.put("device_id", this.sharedData.getThisDevice().getDevicePubKeyStr());
                generatedRequest.put("result", resultMessage);
                generatedRequest.put("audit_start", this.sharedData.getCurrentSession().getCurrentUTicketId());
                generatedRequest.put("associated_plaintext_data", this.sharedData.getCurrentSession().getAssociatedPlaintextData());
                generatedRequest.put("ciphertextData", this.sharedData.getCurrentSession().getCiphertextData());
                generatedRequest.put("gcmAuthenticationTagData", this.sharedData.getCurrentSession().getGcmAuthenticationTagData());
                generatedRequest.put("ivCmd", this.sharedData.getCurrentSession().getIvCmd());
            } else {
                generatedRequest.put("r_ticket_type", RTicket.TYPE_DATA_RTOKEN);
                generatedRequest.put("device_id", this.sharedData.getThisDevice().getDevicePubKeyStr());
                generatedRequest.put("result", resultMessage);
            }

            String generatedRTicketJson = this.msgGenerator.generateXxxRTicket(generatedRequest);

            // End Process Measurement
            this.measureHelper.measureRecvMsgPerfTime("deviceRecvCmdAndSendData");

            // Start Comm Measurement
            if (!this.sharedData.getThisDevice().getDeviceName().equals("iotDevice")) {
                this.measureHelper.measureCommPerfStart();
            }

            // [STAGE: (S)]
            this.msgSender.sendXxxMessage(
                    Message.MESSAGE_VERIFY_AND_EXECUTE,
                    RTicket.MESSAGE_TYPE,
                    generatedRTicketJson
            );
        } catch (Exception e) {
            // pragma: no cover -> Shouldn't Reach Here
            throw new RuntimeException("Shouldn't Reach Here", e);
        }

        // Manually Finish Simulated Comm
        SimpleLogger.simpleLog("debug","+ " + this.sharedData.getThisDevice().getDeviceName() + " manually finish PS~~ (device)");

        if (Environment.COMMUNICATION_CHANNEL.equals("SIMULATED")) {
            msgSender.completeSimulatedComm();
        }
    }

    public void _holderRecvData(RTicket receivedRTicket) {
        try {
            // [STAGE: (R)(VR)]

            // Query Corresponding UTicket
            // [STAGE: (VL)(L)]
            String device_id = receivedRTicket.getDeviceId();
            String storedUTicketJson = this.sharedData.getDeviceTable().get(device_id).getDeviceUTicketForOwner();

            SimpleLogger.simpleLog("debug","Corresponding UTicket: " + storedUTicketJson);

            // [STAGE: (VR)]
            UTicket storedUTicket = this.msgVerifier._classifyUTicketIsDefinedType(storedUTicketJson);

            // [STAGE: (VRT)]
            this.msgVerifier.verifyUTicketHasExecutedThroughRTicket(receivedRTicket, storedUTicket, null);

            // [STAGE: (VTK)]
            // [STAGE: (E)]
            this.executor.executeXxxRTicket(receivedRTicket, "holderOrDevice");
            this.sharedData.setResultMessage("-> SUCCESS: VERIFY_UT_HAS_EXECUTED");

            // [STAGE: (C)]
            this.executor.changeState(ThisDevice.STATE_DEVICE_WAIT_FOR_CMD);

            SimpleLogger.simpleLog("debug","resultMessage = " + this.sharedData.getResultMessage());

        } catch (NoSuchElementException e) {  // pragma: no cover -> FAILURE: (VL)
            String error = "FAILURE: (VL)";
            this.sharedData.setResultMessage(error);
            throw new RuntimeException(error, e);

        } catch (RuntimeException e) { // FAILURE: (VR)(VRT)(VTK)
            this.sharedData.setResultMessage(e.getMessage());

        } catch (Exception e) { // pragma: no cover -> Shouldn't Reach Here
            throw new RuntimeException("Shouldn't Reach Here", e);
        } finally {
            // End Process Measurement
            this.measureHelper.measureRecvMsgPerfTime("holderRecvData");

            // Start Comm Measurement
            if (!this.sharedData.getThisDevice().getDeviceName().equals("iotDevice")) {
                this.measureHelper.measureCommPerfStart();
            }
        }

        // Manually Finish Simulated/Bluetooth Comm
        SimpleLogger.simpleLog("debug","+ " + this.sharedData.getThisDevice().getDeviceName() + " manually finish PS~~ (holder)");

        if (Environment.COMMUNICATION_CHANNEL.equals("SIMULATED")) {
            this.msgSender.completeSimulatedComm();
//        } else if (Environment.COMMUNICATION_CHANNEL.equals("BLUETOOTH")) {
//            this.msgSender.completeBluetoothComm();
        }
    }
}